### PR TITLE
Update / Fixes for MultiSig ownership and Authorization issuance

### PR DESCRIPTION
**Pull request by adamdossa**
This PR contains:

**Authorizations:**

  - all authorizations are now `authorized_by` an identity (DID), not a signatory (account or DID)
  - corresponding changes to `AuthorizationData`

**MultiSigs:**

`Signers`: the keys attached to the MultiSig as signers
`Creator Identity`: the identity that created the MultiSig - this is fixed and can't be modified
`Acting Identity`: the identity that the MultiSig acts through, by virtue of being attached as a Signing / Master Key.

`Signers => MultiSig` is mapped through `KeyToMultiSig`
`MultiSig => Creator Identity` is mapped through `MultiSigToIdentity`
`MultiSig => Acting Identity` is mapped through `KeyToIdentityIds` (`None` if MultiSig is not a Signing / Master Key)

This has the consequence that Signers can not send / receive POLYX (since they are not directly associated with an identity). Note that in Aldebaran, it is possible for keys to hold funds, but they can't pay for transactions.

MultiSigs can hold funds, once they have been attached as a Signing / Master Key to an identity. If they are removed from the identity they can continue to hold funds, as per normal external key behaviour, but these funds are locked until the key is reattached to a CDD'ed identity.

All transactions have a single payer, which is established when the transaction is first submitted. For a transaction executed via a MultiSig this will be the Creator Identity derived from the external Signer account that submits the extrinsic.

When Signers are approving / proposing a MultiSig transaction, we have:
  - CDD check done on `Creator Identity`
  - Payment made by `Creator Identity`
  - No `current_identity` set

When a MultiSig executes a proposal, we have:
  - CDD check done on `Creator Identity` (through the SignedExtension)
  - CDD check done on `Acting Identity` (on dispatch)
  - Payment made by `Creator Identity` (bug - currently Protocol Fee Payment made by MultiSig)
  - `current_identity` set to `Acting Identity`

**Bug Fixes:**
  - Previously a MS signer could execute a tx (e.g. asset.registerTicker) and assume the identity of the MS creator.
  - When rotating master keys, new key was not linked to DID.
  - Fix bug where the same signing key can be attached twice with different permissions to the same identity.

### CODE REVIEW

**Comment by adamdossa** [comment]
@satyamakgec I am getting a weird testing error since merging w/ `develop` - did you see anything like this when doing the rc.4 upgrade?

```
adamdossa@Adams-MBP Polymesh % ./scripts/test.sh

warning: unused variable: `who`
   --> pallets/transaction-payment/src/lib.rs:499:19
    |
499 |         let (tip, who, imbalance, fee) = pre;
    |                   ^^^ help: consider prefixing with an underscore: `_who`
    |
    = note: `#[warn(unused_variables)]` on by default

   Compiling pallet-identity v0.1.0 (/Users/adamdossa/Development/Polymath/Polymesh/pallets/identity)
   Compiling substrate-test-runtime v2.0.0-rc3 (https://github.com/paritytech/substrate?rev=c771821cae2dcb5a8808a19fae8122c0b9ae8499#c771821c)
error: couldn't read /Users/adamdossa/Development/Polymath/Polymesh/target/debug/build/substrate-test-runtime-498607dc68a08276/out/wasm_binary.rs: No such file or directory (os error 2)
  --> /Users/adamdossa/.cargo/git/checkouts/substrate-7e08433d4c370a21/c771821/test-utils/runtime/src/lib.rs:62:1
   |
62 | include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `substrate-test-runtime`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
warning: unused import: `Dispatchable`
   --> pallets/identity/src/lib.rs:117:42
    |
117 |         AccountIdConversion, CheckedAdd, Dispatchable, Hash, IdentifyAccount, SaturatedConversion,
    |                                          ^^^^^^^^^^^^
    |
    = note: `#[warn(unused_imports)]` on by default

warning: unused variable: `sender`
   --> pallets/identity/src/lib.rs:568:17
    |
568 |             let sender = ensure_signed(origin)?;
    |                 ^^^^^^ help: consider prefixing with an underscore: `_sender`
    |
    = note: `#[warn(unused_variables)]` on by default

error: build failed
```

**Comment by maxsam4** [comment]
@adamdossa For a faster test compilation, the test script does not compile the WASM binary. It assumes that you already have it compiled. Even an old version is fine since tests run natively. You can generate a WASM but running tests normally once and then you can use the script. To run tests normally, execute `cargo test --package polymesh-runtime-common --features default_identity`.

The test script basically uses the env var `SKIP_WASM_BUILD=true` to speed up the process. Your previous wasm build became invalid after the substrate and hence you are getting that error.

File: pallets/identity/src/lib.rs [hunk 1]
```diff
@@ -1174,72 +1159,57 @@ impl<T: Trait> Module<T> {
...
+            &Signatory::Identity(target_did),
+            ProtocolOp::IdentityAddSigningItemsWithAuthorization,
+        )?;
         if let Signatory::Account(key) = &signer {
-            if !Self::can_key_be_linked_to_did(key, SignatoryType::External) {
-                ensure!(
-                    Self::get_identity(key) == Some(identity_data_to_join.target_did),
-                    Error::<T>::AlreadyLinked
-                )
-            } else {
```
Comment (maxsam4) [inline pallets/identity/src/lib.rs:1205]: I can't remember why but there might be a reason why we added this condition. It basically allows an existing key to accept the authorization without errors.

**Comment by adamdossa** [comment]
@satyamakgec sorry - another question - some of the UTs are failing with a "not implemented" error - is this something you saw before (see circleci logs)?

**Comment by maxsam4** [comment]
> @satyamakgec sorry - another question - some of the UTs are failing with a "not implemented" error - is this something you saw before (see circleci logs)?

That'll happen when your code reached an `unimplemented!()` block. Your tests are somehow hitting the new multisig functions you added like `is_signer()` (or an old function that wasn't being triggered before) that have been set as `unimplemented!()` in the test mock.

**Comment by adamdossa** [comment]
> @adamdossa For a faster test compilation, the test script does not compile the WASM binary. It assumes that you already have it compiled. Even an old version is fine since tests run natively. You can generate a WASM but running tests normally once and then you can use the script. To run tests normally, execute `cargo test --package polymesh-runtime-common --features default_identity`.
> 
> The test script basically uses the env var `SKIP_WASM_BUILD=true` to speed up the process. Your previous wasm build became invalid after the substrate and hence you are getting that error.

That makes sense - I did compile before running tests but it looks like it is looking for a debug wasm build and I only had a release build.

**Comment by satyamakgec** [comment]
@adamdossa Sorry, I missed all the pings from that PR.

**Comment by adamdossa** [comment]
@maxsam4 @satyamakgec remaining TODO has been moved to a separate JIRA - to correctly use the set payer_context when paying protocol fees (rather than re-specifying it every time we call `ProtocolFee::charge_fee`).

File: pallets/common/src/traits/multisig.rs [hunk 2]
```diff
@@ -42,4 +43,9 @@ pub trait MultiSigSubTrait<AccountId> {
     /// # Arguments
     /// * `account` - AccountId to check
     fn is_multisig(account: &AccountId) -> bool;
+    /// Helper function that returns if a key is a multisig signer
+    fn is_signer(key: &AccountId) -> bool;
+
+    /// Helper function that returns if a key is a multisig signer
+    fn did_of_signer(key: &AccountId) -> IdentityId;
```
Comment (maxsam4) [inline pallets/common/src/traits/multisig.rs:50]: The comment needs updating.

File: pallets/multisig/src/lib.rs [hunk 3]
```diff
@@ -1111,6 +1106,17 @@ impl<T: Trait> MultiSigSubTrait<T::AccountId> for Module<T> {
...
-        <NumberOfSigners<T>>::contains_key(account)
+        <MultiSigToIdentity<T>>::contains_key(account)
+    }
+
+    /// Helper function that returns if a key is a multisig signer
+    fn is_signer(key: &T::AccountId) -> bool {
+        <KeyToMultiSig<T>>::contains_key(key)
+    }
+
+    /// Helper function that returns if a key is a multisig signer
```
Comment (maxsam4) [inline pallets/multisig/src/lib.rs:1117]: Comments needs updating.

File: pallets/common/src/traits/multisig.rs (hunk 2 above)
Comment (adamdossa) [reply pallets/common/src/traits/multisig.rs]: This function was removed as it is no longer needed

File: pallets/multisig/src/lib.rs (hunk 3 above)
Comment (adamdossa) [reply pallets/multisig/src/lib.rs]: This function was removed as it is no longer needed

**Pull request merged**